### DISCUSSION TITLE
Minimum requirement for a circuit

### DISCUSSION THREAD

**Discussion by fosgate29** [Consensys/gnark#110] [Q&A] [answered]
Hi, I build a pretty simple cirtuit, but it won't work. I get a `runtime error: index out of range [0] with length 0`. 
My struct is 
```
type circuitv5 struct {
	AcceptedQuote frontend.Variable `gnark:",public"` 
	QuoteFromA    frontend.Variable `gnark:",private"` 
}

```
and my circuit has only this line:
`cs.AssertIsEqual(circuit.AcceptedQuote, circuit.QuoteFromA)`

When I setup, it fails with the error message above.

But if I add a multiplication, it works:
`cs.Mul(circuit.AcceptedQuote, circuit.QuoteFromA)`.

So if there is **Mul** before the Assert, it works. Is it a must to have a computation before? Or am I doing something wrong or missunderstood it?

Thanks.
here is my repo in case you would like to run it: https://github.com/fosgate29/financialcircuit/blob/master/bond5.go

**Comment by gbotrel** [comment, marked as answer]
Hi, indeed this popped up before, and this should throw a clean error message ( #112 ). 

A circuit that has no computational constraints make little sense (do you have something we missed in mind? ) --> it means the proof is stating something like `x == y` and to verify the proof, one must supply... `y`. 

Note that some `assertions` (like [`AssertIsLessOrEqual`](https://pkg.go.dev/github.com/consensys/gnark@v0.4.0/frontend#ConstraintSystem.AssertIsLessOrEqual)) will add computational constraints. `AssertIsEqual` does not.

**Reply by fosgate29** [reply]
Thank you. I will give it another try. Thanks.

**Reply by fosgate29** [reply]
Actually we were trying to build this circuit: 
```
type bondCircuitv struct {
	AcceptedQuote frontend.Variable `gnark:",public"`  
	QuoteFromA     frontend.Variable `gnark:",private"` 
	QuoteFromB    frontend.Variable `gnark:",private"` 
	QuoteFromC    frontend.Variable `gnark:",private"`
}
```
And with this logic to get the best quote:

```
	if(QuoteFromA < QuoteFromB){
		if(QuoteFromA < QuoteC){
			Result = QuoteFromA
		}
		else{
			Result = QuoteFromC
		}
	}
	else if(QuoteFromB < QuoteFromC){
		Result = QuoteFromB
	}
	else{
		Result = QuoteFromC
	} 
	
	cs.AssertIsEqual(cs.AcceptedQuote, Result)
```

But we are not able to compare QuoteFromA, QuoteFromB and QuoteFromC and get as a result the best quote.

**Reply by fosgate29** [reply]
I built this one:
```
type bondCircuitv5 struct {
	//Bid 92.63
	AcceptedQuote frontend.Variable `gnark:",public"` // 92
	QuoteFromA  frontend.Variable `gnark:",private"` // 92
	QuoteFromB  frontend.Variable `gnark:",private"` // 93
	QuoteFromC  frontend.Variable `gnark:",private"` // 95
	WinnerQuote frontend.Variable `gnark:",private"` // 92
	Quote1      frontend.Variable `gnark:",private"` // 93
	Quote2      frontend.Variable `gnark:",private"` // 95
}

func (circuit *bondCircuitv5) Define(curveID ecc.ID, cs *frontend.ConstraintSystem) error {

	//Make sure Winner Quote is the smallest one
	cs.AssertIsLessOrEqual(circuit.WinnerQuote, circuit.Quote1)
	cs.AssertIsLessOrEqual(circuit.WinnerQuote, circuit.Quote2)
	cs.AssertIsEqual(circuit.WinnerQuote, circuit.AcceptedQuote)

	//If winner quote is from A, B or C, one of the subtraction is going to return zero
	subA := cs.Sub(circuit.QuoteFromA, circuit.WinnerQuote)
	subB := cs.Sub(circuit.QuoteFromB, circuit.WinnerQuote)
	subC := cs.Sub(circuit.QuoteFromC, circuit.WinnerQuote)

	outputA := cs.IsZero(subA, curveID) // 1 - iszero - true
	outputB := cs.IsZero(subB, curveID) // 0 false
	outputC := cs.IsZero(subC, curveID) // 0 false

	// outputA || outputB || outputC == 1
	result_temp := cs.Or(outputA, outputB)
	result := cs.Or(result_temp, outputC)

	one := cs.Constant(1)
	cs.AssertIsEqual(result, one) //
```

I need to add a Signature to make sure A, B and C knows their quotes are inside the circuit. PubKey from A, B and C is going to be public parameters. 

I post here when it is ready.
